### PR TITLE
Fix incorrect simplification of CASE with constant boolean results

### PR DESCRIPTION
**Pull request by martint**
Expressions such as:
```
   CASE
      WHEN <e1> THEN <boolean>
      WHEN <e2> THEN <boolean>
      ...
      ELSE <boolean>
   END
```
were being incorrectly optimized. The terms in the rewritten boolean expression need to reflect the fact that all the preceding conditions should've evaluated to false in order for a given condition to evaluate to true. The optimization was ignoring them, which could result in incorrect results when the conditions are not exclusive.

For example:
```
   CASE
      WHEN a OR b THEN false
      WHEN a OR c THEN true
   END
```
Previously, it was being rewritten as:
```
   (a OR c) ≡ true
```
However, if `a` is true, then the first branch of the CASE should be evaluated and the expression should return false.

Fixes https://github.com/trinodb/trino/issues/23787

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
## Section
* Fix incorrect results for certain CASE expressions that return boolean results. ({issue}`23787`)
```

### CODE REVIEW

**Review by losipiuk — APPROVED** [review]

File: core/trino-main/src/test/java/io/trino/sql/ir/optimizer/TestSimplifyRedundantCase.java [hunk 1]
```diff
@@ -92,7 +111,54 @@ void test()
...
+                                        new Comparison(IDENTICAL, new Reference(BOOLEAN, "a3"), TRUE),
+                                        IrExpressions.not(PLANNER_CONTEXT.getMetadata(), new Comparison(IDENTICAL, new Reference(BOOLEAN, "a4"), TRUE))))));
+
+        assertThat(optimize(
+                new Case(
+                        ImmutableList.of(
+                                new WhenClause(new Reference(BOOLEAN, "a1"), TRUE),
+                                new WhenClause(new Reference(BOOLEAN, "a2"), FALSE),
+                                new WhenClause(new Reference(BOOLEAN, "a3"), FALSE)),
+                        FALSE)))
```
Comment (Praveen2112) [inline core/trino-main/src/test/java/io/trino/sql/ir/optimizer/TestSimplifyRedundantCase.java:160]: What if all the when clause return `false` and default return `false` -> we could convert them to `FALSE` right ? Would we handle them here or at different rule ?

File: core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyRedundantCase.java [hunk 2]
```diff
@@ -65,21 +67,58 @@ public Optional<Expression> apply(Expression expression, Session session, Map<Sy
...
+        //   WHEN a4 THEN false
+        //   WHEN a5 THEN true
+        //   WHEN a6 THEN false
+        //   ELSE true
+        // END
+        //
+        // can be transformed to:
+        //     (a1 ≢ true AND a2 ≢ true AND a3 ≡ true) OR
+        //     (a1 ≢ true AND a2 ≢ true AND a3 ≢ true AND a4 ≢ true AND a5 ≡ true) OR
+        //     (a1 ≢ true AND a2 ≢ true AND a3 ≢ true AND a4 ≢ true AND a5 ≢ true AND a6 ≢ true)
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyRedundantCase.java:89]: It's just factoring out common terms.

File: core/trino-main/src/test/java/io/trino/sql/ir/optimizer/TestSimplifyRedundantCase.java (hunk 1 above)
Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/ir/optimizer/TestSimplifyRedundantCase.java:160]: Good catch. There was supposed to be a test for that at the top of this method, but I realize I made a copy/paste mistake.

Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/ir/optimizer/TestSimplifyRedundantCase.java:160]: Updated.

File: core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyRedundantCase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyRedundantCase.java:90]: To be more precise, it relies on the following equivalences:

* `(a AND b) OR (a AND c)` <=> `a AND (b OR c)`
* `a OR (!a AND b)` <=> `a OR b`

**Pull request merged**